### PR TITLE
Iaas -  add anitmalware extension

### LINKED ISSUE #230: New-CmAzIaasVm support for Defender/Antimalware extension
Add the anti-malware extension to our VM provisioning ARM template/script. An example JSON can be found [as a quickstart here](https://github.com/Azure/azure-quickstart-templates/blob/master/anti-malware-extension-windows-vm/). In addition, [this blog post](https://charbelnemnom.com/how-to-update-microsoft-antimalware-policies-for-azure-virtual-machines-with-powershell-arm-json-azuresecurity/) has some useful context, while [this one](https://www.rdsgurus.com/azure-resource-manager-and-json-templates-to-deploy-rds-in-azure-iaas-part-8-defender-bginfo/) seems to show non-preview API versions being used.

Suggested settings file structure looks something like this (note this would sit in the objects within the `virtualmachines` array:

``` yaml
 - name: myVm
   size: DS3v2_16
   antimalware:
     enable: true
     realtime:
       enabled: true
       excludePaths:
       - "path1"
       - "path2"
       excludeExtensions:
       - ".mdf"
       - ".ldf"
       excludeProcesses:
       - "something.exe"
       - "somethingelse.exe"
     scheduled:
       enabled: true
       scanType": "Quick"
       day: 7
       time: 120
```

### PR DESCRIPTION
**Pull request by deepend-dev**
resolves #230

### CODE REVIEW

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json [hunk 1]
```diff
@@ -430,7 +473,9 @@
                             "dependsOn": [
                                 "[resourceId(parameters('VM').resourceGroupName, 'Microsoft.Compute/virtualMachines/extensions', parameters('VM').fullName, variables('deploymentDetails').dependencyAgentExtensionType)]",
                                 "[resourceId(parameters('VM').resourceGroupName, 'Microsoft.Compute/virtualMachines/extensions', parameters('VM').fullName, variables('deploymentDetails').monitoringAgentExtensionName)]",
-                                "[resourceId(parameters('VM').resourceGroupName, 'Microsoft.Compute/virtualMachines/extensions', parameters('VM').fullName, variables('deploymentDetails').dscExtensionName)]"
+                                "[resourceId(parameters('VM').resourceGroupName, 'Microsoft.Compute/virtualMachines/extensions', parameters('VM').fullName, variables('deploymentDetails').dscExtensionName)]",
+                                "[resourceId(parameters('VM').resourceGroupName, 'Microsoft.Compute/virtualMachines/extensions', parameters('VM').fullName, 'Antimalware')]",
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json:477]: Can we make 'Antimalware' and any other repeated hard coded strings into variables please

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 2]
```diff
@@ -247,12 +248,45 @@
...
+							paths      = $virtualMachine.antimalware.exclusions.paths -join (';');
+							extensions = $virtualMachine.antimalware.exclusions.extensions -join (';');
+							processes  = $virtualMachine.antimalware.exclusions.processes -join (';')
+						}
+
+						if ($virtualMachine.antimalware.schedule) {
+							$virtualMachine.antimalware.schedule.scanType ??= "Quick"
+							$virtualMachine.antimalware.schedule.day = $localConfigs.days.$($virtualMachine.antimalware.schedule.day)
+							$virtualMachine.antimalware.schedule.time = $virtualMachine.antimalware.schedule.time.toString()
+							$virtualMachine.antimalware.schedule.isEnabled = "true"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:270]: Isnt this a boolean?

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 3]
```diff
@@ -247,12 +248,45 @@
...
+						}
+						else {
+							$virtualMachine.antimalware.schedule = @{isEnabled = "false"}
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:273]: @{ isEnabled = "false" }

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 4]
```diff
@@ -247,12 +248,45 @@
...
+						}
+
+						$virtualMachine.antimalware.realtimeProtection = $virtualMachine.antimalware.realtimeProtection ? "true" : "false"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:276]: Isnt this a boolean?

File: Cloudmarque.Azure/Resources/Project/virtualmachines.yml [hunk 5]
```diff
@@ -90,6 +90,22 @@ groups:
               vnet: Iaas-Test-Spoke1-Vnet1
           ipAddress:                                              # Leave blank for dynamic ip allocations
           subnetName: Iaas-Test-Spoke1-Vnet1-Snet-hpcman          # Subnet belonging to the virtual network
+        antimalware:
+          enable: true
+          realTimeProtection: true
+          exclusions:
+            paths:
+            - "D:\\somelogs"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/virtualmachines.yml:98]: No need for " in strings

File: Cloudmarque.Azure/Resources/Project/virtualmachines.yml [hunk 6]
```diff
@@ -315,9 +331,26 @@ groups:
...
           ipAddress:                                              # Leave blank for dynamic ip allocations
-          subnetName: Iaas-Test-Spoke1-Vnet2-Snet-hpcman          # Subnet belonging to the virtual network
+          subnetName: Iaas-Test-Int-Vnet1-Snet-ext1               # Subnet belonging to the virtual network
+        vulnerabilityScan: true
+        antimalware:
+          enable: true
+          realTimeProtection: true
+          exclusions:
+            paths:
+            - "D:\\somelogs"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/virtualmachines.yml:343]: No need for " in strings

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 7]
```diff
@@ -665,6 +671,95 @@
...
+															"$id": "#/groups/items/allOf/virtualMachines/items/allOf/antimalware/enable",
+															"type": "boolean",
+															"title": "Enable",
+															"description": "Status of Antimalware service on virtual machine."
+														},
+														"realTimeProtection": {
+															"$id": "#/groups/items/allOf/virtualMachines/items/allOf/antimalware/realTimeProtection",
+															"type": ["boolean", "null"],
+															"title": "RealTimeProtection",
+															"description": "Status of Antimalware service realTime Protection Feature."
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:694]: lowercase p in protection and f in feature

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 8]
```diff
@@ -665,6 +671,95 @@
...
+														},
+														"exclusions": {
+															"$id": "#/groups/items/allOf/virtualMachines/items/allOf/antimalware/exclusions",
+															"type": ["object", "null"],
+															"title": "Exclusions",
+															"description": "Container of exclusions from scanner.",
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:700]: Add vulnerability before scanner where applicable

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json [hunk 9]
```diff
@@ -665,6 +671,95 @@
...
+																	"description": "Paths to exclude from scanner.",
+																	"enum": ["Quick", "Full"]
+																},
+																"day": {
+																	"$id": "#/groups/items/allOf/virtualMachines/items/allOf/antimalware/schedule/extensions",
+																	"type": ["string", "null"],
+																	"title": "Day",
+																	"description": "Day of the week for scheduled run.",
+																	"enum": [
+																		"sunday", "monday", "tuesday", "wednesday", "thrusday", "friday", "saturday"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:747]: Should these be Pascal case?

File: azure-integration-pipelines.yml [hunk 10]
```diff
@@ -132,31 +132,31 @@ stages:
...
+  #         ./Initialise-Cloudmarque.ps1 `
+  #           -ProjectDirectory "$(Build.SourcesDirectory)" `
+  #           -Environment "$(environment)" `
+  #           -BuildNumber "$(Build.BuildNumber)" `
+  #           -RepositoryName "$(Build.Repository.Name)" `
+  #           -DefinitionName "$(Build.DefinitionName)" `
+  #           -DefinitionVersion "$(Build.DefinitionVersion)"
+
+  #         New-CmAzIaasBastionHost -SettingsFile "$(Build.SourcesDirectory)/$(moduleName)/Resources/Project/bastionHost.yml" -Verbose
+  #         New-CmAzIaasVpnGw -SettingsFile "$(Build.SourcesDirectory)/$(moduleName)/Resources/Project/vpngw.yml" -Verbose
```
Comment (tm-luke) [inline azure-integration-pipelines.yml:159]: Remove this change

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 11]
```diff
@@ -247,12 +248,45 @@
...
 
 						Write-Verbose "TimeZone settings found."
-						$virtualMachine.timeZone = $allowedTimeZones.timeZones.contains($virtualMachine.timeZone) ? $virtualMachine.timeZone : (Write-Error "Please provide valid Windows timezone format...")
+						$virtualMachine.timeZone = $localConfigs.timeZones.contains($virtualMachine.timeZone) ? $virtualMachine.timeZone : (Write-Error "Please provide valid Windows timezone format...")
 					}
 					else {
 						$virtualMachine.timeZone = "UTC"
 					}
 
+					if ($virtualMachine.antimalware) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:257]: This check seems a little redundant as the json is enforcing the required settings here

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 [hunk 12]
```diff
@@ -247,12 +248,45 @@
...
+
+						Write-Verbose "Antimalware settings found."
+						$virtualMachine.antimalware.exclusions = @{
+							paths      = $virtualMachine.antimalware.exclusions.paths -join (';');
+							extensions = $virtualMachine.antimalware.exclusions.extensions -join (';');
+							processes  = $virtualMachine.antimalware.exclusions.processes -join (';')
+						}
+
+						if ($virtualMachine.antimalware.schedule) {
```
Comment (tm-luke) [inline Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:266]: This check seems a little redundant as the json is enforcing the required settings here

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json [hunk 13]
```diff
@@ -422,6 +422,48 @@
                                 "settings": "[variables('deploymentDetails').dscSettings]"
                             }
                         },
+                        {
+                            "condition": "[and(parameters('VM').antimalware.enable, equals(parameters('VM').osDisk.osType, 'windows'))]",
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json:426]: replaced with variables('isWindows')

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json (hunk 1 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json:477]: none of them are repeated as such.. and they are same for linux and windows just like diskencryption

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 2 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:270]: nope

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 11 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:257]: nope.. we have to convert array to semicolon sepeated string

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json (hunk 1 above)
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json:477]: Not sure if i understand, theres three instances of the word that i can see in the arm template, may as well use a variable instead

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 7 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:694]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 12 above)
Comment (tm-luke) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:266]: The json schema should at least guarantee this value @{ isEnabled = "false" }, so theres probably no need for the else statement.

File: Cloudmarque.Azure/Resources/Project/Integration/virtualmachines.yml [hunk 14]
```diff
@@ -90,6 +90,22 @@ groups:
               vnet: Iaas-Test-Spoke1-Vnet1
           ipAddress:                                              # Leave blank for dynamic ip allocations
           subnetName: Iaas-Test-Spoke1-Vnet1-Snet-hpcman          # Subnet belonging to the virtual network
+        antimalware:
+          enable: true
+          realTimeProtection: true
+          exclusions:
+            paths:
+            - "D:\\somelogs"
```
Comment (tm-luke) [inline Cloudmarque.Azure/Resources/Project/Integration/virtualmachines.yml:98]: Can remove " for strings that contain letters, this also applicable for all other settings

File: Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json (hunk 9 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Schema/New-CmAzIaasVm.Schema.json:747]: not hard coded any where in azure.. as per our naming single string in yml are lowercase

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json [hunk 15]
```diff
@@ -422,6 +422,49 @@
...
+                                "typeHandlerVersion": "1.3",
+                                "autoUpgradeMinorVersion": true,
+                                "settings": {
+                                    "AntimalwareEnabled": true,
+                                    "Exclusions": {
+                                        "Paths": "[parameters('VM').antimalware.exclusions.paths]",
+                                        "Extensions": "[parameters('VM').antimalware.exclusions.extensions]",
+                                        "Processes": "[parameters('VM').antimalware.exclusions.processes]"
+                                    },
+                                    "RealtimeProtectionEnabled": "[parameters('VM').antimalware.realtimeProtection]",
```
[the remark this answers is not in the document]
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json]: done

**Review by tm-luke — CHANGES_REQUESTED** [review]

File: azure-integration-pipelines.yml (hunk 10 above)
Comment (deepend-dev) [reply azure-integration-pipelines.yml]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1 (hunk 12 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.ps1:266]: no it does not.. as notice its a string not a boolean which we get if schedule property is defined or not

File: Cloudmarque.Azure/Resources/Project/Integration/virtualmachines.yml (hunk 14 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Resources/Project/Integration/virtualmachines.yml:98]: done

File: Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json (hunk 1 above)
Comment (deepend-dev) [reply Cloudmarque.Azure/Public/Deploy/IaaS/vm/New-CmAzIaasVm.json]: Done it but not sure if we have done this in other arm templates.

**Review by tm-luke — DISMISSED** [review]

**Review by tm-luke — APPROVED** [review]

**Pull request merged**